File: 06-social-media-scripts/docs/research/ig-tools-bench/e2e.py
"""E2E real: últimos N reels de una cuenta CON métricas + caption + url de video."""
import os, time, json
import instaloader

N = 10
ACCOUNTS = ["nasa", "openai"]
L = instaloader.Instaloader(quiet=True, save_metadata=False, download_comments=False)
sid = os.environ.get("IG_SESSIONID", "").strip()
if sid:
    L.context._session.cookies.set("sessionid", sid, domain=".instagram.com")
    try:
        L.context.username = L.test_login()
    except Exception as e:
        print("[warn] test_login:", e)

out = []
for acc in ACCOUNTS:
    t0 = time.perf_counter()
    row = {"account": acc}
    try:
        prof = instaloader.Profile.from_username(L.context, acc)
        row["t_listing_s"] = round(time.perf_counter() - t0, 2)
        t1 = time.perf_counter()
        full, scanned = [], 0
        for p in prof.get_posts():
            scanned += 1
            if p.is_video:
                full.append({"shortcode": p.shortcode, "likes": p.likes, "comments": p.comments,
                             "play_count": p.video_view_count, "caption": p.caption,
                             "duration": p.video_duration, "video_url": p.video_url,
                             "music": None})
            if len(full) >= N or scanned >= 25:
                break
        row["t_enrich_s"] = round(time.perf_counter() - t1, 2)
        row["t_total_s"] = round(time.perf_counter() - t0, 2)
        row["reels"] = len(full)
        row["posts_scanned"] = scanned
        row["with_caption"] = sum(1 for r in full if r["caption"])
        row["with_video"] = sum(1 for r in full if r["video_url"])
        row["with_views"] = sum(1 for r in full if r["play_count"] is not None)
        row["with_music"] = sum(1 for r in full if r["music"])
        row["s_per_reel"] = round(row["t_total_s"] / max(len(full), 1), 2)
        row["status"] = "ok"
    except Exception as e:
        row["status"] = "error"
        row["error"] = f"{type(e).__name__}: {str(e)[:180]}"
        row["t_total_s"] = round(time.perf_counter() - t0, 2)
    print(json.dumps(row, ensure_ascii=False, default=str)[:400])
    out.append(row)

json.dump({"tool": "instaloader", "results": out}, open("results/e2e_instaloader.json", "w"),
          indent=1, ensure_ascii=False, default=str)

File: 06-social-media-scripts/docs/research/ig-tools-bench/e2e.mjs
/** E2E real: últimos N reels de una cuenta CON métricas + caption + url de video. */
import { igApi } from 'insta-fetcher';
import fs from 'node:fs';

const N = 10;
const ACCOUNTS = ['nasa', 'openai'];
const PARALLEL = Number(process.env.PARALLEL ?? 1);
const ig = new igApi(`sessionid=${(process.env.IG_SESSIONID ?? '').trim()};`);
const sec = (t) => Number(((performance.now() - t) / 1000).toFixed(2));

async function mapLimit(items, limit, fn) {
  const out = new Array(items.length);
  let i = 0;
  await Promise.all(Array.from({ length: Math.min(limit, items.length) }, async () => {
    while (i < items.length) { const k = i++; out[k] = await fn(items[k], k); }
  }));
  return out;
}

const out = [];
for (const acc of ACCOUNTS) {
  const row = { account: acc, parallel: PARALLEL };
  const t0 = performance.now();
  try {
    // paso 1: listado (métricas)
    let t = performance.now();
    const reel = await ig.fetchUserReel(acc, null, N);
    const media = (reel?.xdt_api__v1__clips__user__connection_v2?.edges ?? [])
      .map(e => e.node.media).slice(0, N);
    row.t_listing_s = sec(t);

    // paso 2: enriquecer cada reel (caption + video url)
    t = performance.now();
    const full = await mapLimit(media, PARALLEL, async (m) => {
      try {
        const raw = await ig.fetchPostByMediaId(m.pk);
        const it = raw?.items?.[0] ?? {};
        return {
          shortcode: m.code, likes: m.like_count, comments: m.comment_count,
          play_count: m.play_count, caption: it.caption?.text ?? null,
          duration: it.video_duration ?? null,
          video_url: it.video_versions?.[0]?.url ?? null,
          music: it.clips_metadata?.music_info?.music_asset_info?.title
              ?? it.clips_metadata?.original_sound_info?.original_audio_title ?? null,
        };
      } catch (e) { return { shortcode: m.code, error: String(e?.message).slice(0, 80) }; }
    });
    row.t_enrich_s = sec(t);
    row.t_total_s = sec(t0);
    row.reels = full.length;
    row.with_caption = full.filter(r => r.caption).length;
    row.with_video = full.filter(r => r.video_url).length;
    row.with_views = full.filter(r => r.play_count != null).length;
    row.with_music = full.filter(r => r.music).length;
    row.s_per_reel = Number((row.t_total_s / Math.max(full.length, 1)).toFixed(2));
    row.status = 'ok';
    row.sample = full[0];
  } catch (e) {
    row.status = 'error'; row.error = `${e?.name}: ${String(e?.message).slice(0, 200)}`;
    row.t_total_s = sec(t0);
  }
  console.log(JSON.stringify({ ...row, sample: undefined }));
  out.push(row);
}
fs.writeFileSync(`results/e2e_instafetcher_p${PARALLEL}.json`,
  JSON.stringify({ tool: 'insta-fetcher', parallel: PARALLEL, results: out }, null, 1));
